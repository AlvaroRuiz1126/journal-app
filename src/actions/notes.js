import Swal from 'sweetalert2';
import { types } from "../components/types/types";
import { db } from "../firebase/firebase-config";
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
//react-journal
//Tarea asincrona ya que se hace la peticion en la base de datos de firebase
export const startNewNote = () => {
    return async(dispatch, getState) => {
        //para grabar en firestore se necesita el uid
        //con la funcion getState se obtiene el state del redux, muy perecido al useSelector
        const uid = getState().auth.uid;
        //console.log(uid);
        const note = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        //con esta instruccion se inserta en la base de datos de firestore
        const doc = await db.collection(`${uid}/journal/notes`).add(note);
        //console.log(doc);
        //esta accion recibe el id que se crea qn firestore que viene en doc.id, y la nota que se le envia
        dispatch(activeNote(doc.id, note));
        dispatch(addNewNote(doc.id, note));
    };
};

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
});

export const startLoadingNotes = (uid) => {
    return async(dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    };
};

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
});

export const startSaveNote = (note) => {
    return async(dispatch, getState) => {
        const uid = getState().auth.uid;
        if(!note.url){
            delete note.url;
        }
        //se le hace un clon al objeto que llega para no modificar el id de la nota cuando se va actualizar
        const noteToFirestore = {...note};
        //elmina la propiedad id del objeto, para eso sirve el delete
        delete noteToFirestore.id;
        //aca en la url del dc usamos el id de la nota que llega ocmo parametro
        //el .update(dato a almacenar) es la funcion para acutalizar
        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
        dispatch(refreshNote(note.id, noteToFirestore));
        Swal.fire('Saved', note.title, 'success');
    };
};

export const refreshNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
});

export const startUploading = (file) => {
    return async(dispatch, getState) => {
        const {active} = getState().notes;
        Swal.fire({
            title: 'Uploadin...',
            text: 'Please wait...',
            allowOutsideClick: false,
            onBeforeOpen: () => {
                Swal.showLoading();
            }
        });
        //console.log(active, file);
        const fileURL = await fileUpload(file);
        //console.log(fileURL);
        active.url = fileURL;

        dispatch(startSaveNote(active));
        Swal.close();
    };
};

export const startDeleting = (id) => {
    return async(dispatch, getState) => {
        const {uid} = getState().auth;
        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id));
    };
};

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
});

export const noteCleaning = () => ({
    type: types.notesLogoutCleaning
});