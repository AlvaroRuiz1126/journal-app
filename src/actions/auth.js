import {types} from './../components/types/types';
import {firebase, googleAuthProvider} from './../firebase/firebase-config';
import { finishLoading, startLoading } from './ui';
import Swal from 'sweetalert2';
import { noteCleaning } from './notes';

export const login = (uid, name) => {
    return {
        type: types.login,
        payload: {
            uid,
            name
        }
    };
};

export const startLoginWithEmailPassword = (email, password) => {
    return (dispatch) => {
        // setTimeout(() => {
        //     dispatch(login(123, 'alvaro'));
        // }, 3500);
        dispatch(startLoading());
        firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({user}) => {
                dispatch(login(user.uid, user.displayName));
                //se debe colocar dentro de la promesa porque si se coloca fuera se ejecuta de manera sincrona con el anterior dispatch
                dispatch(finishLoading());
            }).catch(e => {
                //console.log(e)
                dispatch(finishLoading());
                Swal.fire('Error', e.message, 'error');
            });
    };
};

export const startRegisterWithEmailPassword = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async({user}) => {
                await user.updateProfile({displayName: name});
                dispatch(login(user.uid, user.displayName));
                //console.log(user);
                //console.log(userCred);
                //dispatch(login(user.uid, user.displayName))
            })
            .catch(e => {
                //console.log(e)
                Swal.fire('Error', e.message, 'error');
            });
    };
};

export const startGoogleLogin = () => {
    //el Popup devuelve unas credenciales de usuario
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
        .then(({user}) => {
            //console.log(userCred);
            dispatch(login(user.uid, user.displayName))
        });
    };
};

export const startLogout = () => {
    //esta funcion es asincrona ebido a que se debe esperar a que la accion se ejecute en firebase
    return async(dispatch) => {
        await firebase.auth().signOut();
        dispatch(logout());
        dispatch(noteCleaning());
    };
};

//esta funcion devuelve un objeto vacio para reestablecer las credenciales
export const logout = () => {
    return {
        type: types.logout
    };
};