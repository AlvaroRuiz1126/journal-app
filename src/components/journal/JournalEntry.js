import React from 'react';
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';

export const JournalEntry = ({id, title, body, date, url}) => {
    //console.log(id, title, body, date, url);
    const dispatch = useDispatch();
    const noteDate = moment(date);
    //console.log(noteDate);

    const hanldeEntryClick = () => {
        dispatch(activeNote(id, {title, body, date, url}));
    };

    return (
        <div 
            className="journal__entry pointer"
            onClick={hanldeEntryClick}
        >
            {
                url &&
                <div 
                    className="journal__entry-picture"
                    style={{
                        backgroundSize: 'cover',
                        backgroundImage: `url(${url})`
                    }}
                ></div>
            }

            <div className="journal__entry-body">
                <p className="journal__entry-title">{title}</p>
                <p className="journal__entry-content">
                    {body}
                </p>
            </div>

            <div className="journal__entry-date-box">
                <span>{noteDate.format('dddd')}</span>
                <h4>{noteDate.format('d')}</h4>
            </div>
        </div>
    );
};
