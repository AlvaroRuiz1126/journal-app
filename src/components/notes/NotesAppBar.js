import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {
    const dispatch = useDispatch();
    const {active} = useSelector(state => state.notes)

    const handleFileChange = (e) => {
        //console.log(e.target);
        const file = e.target.files[0];

        if(file){
            dispatch(startUploading(file));
        }
    };

    const handlePictureClick = () => {
        //console.log('Click');
        //esta linea simula elevento del clic con el id de fileSelector
        document.querySelector('#fileSelector').click();
    };

    const handleSave = () => {
        dispatch(startSaveNote(active));
    };

    return (
        <div className="notes__appbar">
            <span>26 de Octubre del 2020</span>

            <input
                id="fileSelector"
                type="file"
                name="file"
                style={{display:"none"}}
                onChange={handleFileChange}
            />

            <div>
                <button 
                    className="btn"
                    onClick={handlePictureClick}
                >
                    Picture
                </button>

                <button 
                    className="btn"
                    onClick={handleSave}
                >
                    Save
                </button>
            </div>
            
        </div>
    );
};
