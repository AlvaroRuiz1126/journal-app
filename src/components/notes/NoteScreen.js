import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';

export const NoteScreen = () => {
    const dispatch = useDispatch();
    const {active} = useSelector(state => state.notes);
    //console.log(active);

    const [values, handleInputChange, reset] = useForm(active);
    //console.log(values);
    const {body, title, id} = values;

    //El useRef recibe una variable mutable que no va redibujar todo el componente si ella cambia
    const activeId = useRef(active.id);

    const handleDelete = () => {
        dispatch(startDeleting(id));
    };

    //permite no crear un bucle infinito con la nota activa
    useEffect(() => {
        if(active.id !== activeId.current){
            reset(active);
            activeId.current = active.id;
        }
    }, [active, reset]);

    useEffect(() => {
        //console.log(values);
        dispatch(activeNote(values.id, {...values}));
    }, [values, dispatch]);

    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">
                <input
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    value={title}
                    name="title"
                    onChange={handleInputChange}
                />

                <textarea
                    placeholder="What happend today?"
                    className="notes__textarea"
                    value={body}
                    name="body"
                    onChange={handleInputChange}
                ></textarea>

                {
                    (active.url) &&
                    <div className="notes__image">
                        <img
                            src={active.url}
                            alt="Imagen"
                        />
                    </div>
                }
            </div>

            <button
                className="btn btn-danger"
                onClick={handleDelete}
            >
                Delete
            </button>
        </div>
    );
};
