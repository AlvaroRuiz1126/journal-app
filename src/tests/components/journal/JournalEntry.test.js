import '@testing-library/jest-dom';
import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { JournalEntry } from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {};
let store = mockStore(initState);
store.dispatch = jest.fn();
const note = {
    id: 1,
    date: 0,
    title: 'Hola',
    body: 'Mundo',
    url: 'https://local.jpg'
}

describe('Pruebas en JournalEntry', () => {
    const wrapper = mount(
        <MemoryRouter>
            <Provider store={store}>
                <JournalEntry {...note} />
            </Provider>
        </MemoryRouter>
    );

    test('Se debe mostrar el ocmponente correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Se debe llamar el activeNote', () => {
        wrapper.find('.journal__entry').prop('onClick')();
        expect(store.dispatch).toHaveBeenLastCalledWith(
            activeNote(note.id, {...note})
        );
    });
});