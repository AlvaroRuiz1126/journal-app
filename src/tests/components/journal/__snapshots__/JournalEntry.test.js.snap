// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en JournalEntry Se debe mostrar el ocmponente correctamente 1`] = `
<div
  className="journal__entry pointer"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://local.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Hola
    </p>
    <p
      className="journal__entry-content"
    >
      Mundo
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      3
    </h4>
  </div>
</div>
`;
